### PR TITLE
[pyright] [core] misc

### PR DESCRIPTION
**Pull request by smackesey**
### Summary & Motivation

Miscellaneous assortment of typing fixes required to pass `pyright` in core `dagster`. This PR contains all of the fixes that did not fit into any more coherent theme:

- Adding/updating assorted annotations
- Facotring out certain callback types into reusable type aliases
- Guaranteeing certain local variables are defined on access
- A few calls to e.g. `check.not_none` to fix type inference

### How I Tested These Changes

Pyright, BK

### CODE REVIEW

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #11339** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/11339" target="_blank">Graphite</a> 
    * **PR #11334** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/11334" target="_blank">Graphite</a> 
      * **PR #11335** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/11335" target="_blank">Graphite</a> 
        * **PR #10900** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/10900" target="_blank">Graphite</a> 
          * **PR #10901** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/10901" target="_blank">Graphite</a> 
            * **PR #10907** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/10907" target="_blank">Graphite</a> 
              * **PR #10992** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/10992" target="_blank">Graphite</a> 
                * **PR #10993** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/10993" target="_blank">Graphite</a> 
                  * **PR #10994** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/10994" target="_blank">Graphite</a> 
                    * **PR #10995** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/10995" target="_blank">Graphite</a> 
                      * **PR #10996** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/10996" target="_blank">Graphite</a> 
                        * **PR #11329** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/11329" target="_blank">Graphite</a> 
                          * **PR #10997** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/10997" target="_blank">Graphite</a> 
                            * **PR #10998** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/10998" target="_blank">Graphite</a> 
                              * **PR #10999** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/10999" target="_blank">Graphite</a> 
                                * **PR #11000** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/11000" target="_blank">Graphite</a> 
                                  * **PR #11001** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/11001" target="_blank">Graphite</a> 
                                    * **PR #11002** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/11002" target="_blank">Graphite</a> 
                                      * **PR #11330** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/11330" target="_blank">Graphite</a> 
                                      * **PR #11038** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/11038" target="_blank">Graphite</a> 
                                        * **PR #10983** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/10983" target="_blank">Graphite</a> 
                                        * **PR #11051** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/11051" target="_blank">Graphite</a> 
                                        * **PR #11362** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/11362" target="_blank">Graphite</a>  👈
      * **PR #11212** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/11212" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/11362?utm_source=stack-comment).

File: python_modules/dagster/dagster/_utils/backcompat.py [hunk 1]
```diff
@@ -14,7 +14,15 @@
 
 
 def canonicalize_backcompat_args(
```
Comment (schrockn) [inline python_modules/dagster/dagster/_utils/backcompat.py:16]: Potentially dangerous, buggy. Do in separate PR

File: python_modules/dagster/dagster/_utils/error.py [hunk 2]
```diff
@@ -82,7 +83,13 @@ def serializable_error_info_from_exc_info(
     # Whether to forward serialized errors thrown from subprocesses
     hoist_user_code_error: Optional[bool] = True,
 ) -> SerializableErrorInfo:
-    _exc_type, e, _tb = exc_info
+    # `sys.exc_info() return Tuple[None, None, None] when there is no exception being processed. We accept this in
```
Comment (schrockn) [inline python_modules/dagster/dagster/_utils/error.py:86]: Separate PR

File: python_modules/dagster/dagster/_check/__init__.py [hunk 3]
```diff
@@ -167,13 +167,15 @@ def is_callable(obj: object, additional_message: Optional[str] = None) -> Callab
...
-    obj: object,
+    obj: T_Type,
     param_name: str,
     superclass: Optional[type] = None,
     additional_message: Optional[str] = None,
-) -> type:
+) -> T_Type:
     if not isinstance(obj, type):
         raise _param_class_mismatch_exception(
             obj, param_name, superclass, False, additional_message
```
Comment (schrockn) [inline python_modules/dagster/dagster/_check/__init__.py:181]: isn't the point here that the input *can* be anything?

File: python_modules/dagster/dagster/_cli/api.py [hunk 4]
```diff
@@ -95,6 +95,8 @@ def _execute_run_command_body(
         cancellation_thread, cancellation_thread_shutdown_event = start_run_cancellation_thread(
             instance, pipeline_run_id
         )
+    else:
```
Comment (schrockn) [inline python_modules/dagster/dagster/_cli/api.py:98]: Changes in this file separate PR

File: python_modules/dagster/dagster/_core/definitions/asset_reconciliation_sensor.py [hunk 5]
```diff
@@ -566,6 +566,9 @@ def determine_asset_partitions_to_reconcile_for_freshness(
             current_data_times = get_current_data_times_for_key(
                 instance_queryer, asset_graph, relevant_upstream_keys, key
             )
+            expected_data_times = get_expected_data_times_for_key(
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/asset_reconciliation_sensor.py:569]: behavior change?

File: python_modules/dagster/dagster/_core/definitions/scoped_resources_builder.py [hunk 6]
```diff
@@ -25,6 +25,9 @@ class Resources:
     provides a workaround.
     """
 
+    def __getattr__(self, name: str) -> Any:
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/scoped_resources_builder.py:28]: changes in this file in their own PR

File: python_modules/dagster/dagster/_core/storage/event_log/sql_event_log.py [hunk 7]
```diff
@@ -969,14 +969,14 @@ def get_logs_for_all_runs_by_log_id(
...
+        for (
+            record_id,
+            json_str,
+        ) in results:
+            try:
                 events[record_id] = deserialize_as(json_str, EventLogEntry)
-        except (seven.JSONDecodeError, check.CheckError):
-            logging.warning("Could not parse event record id `%s`.", record_id)
+            except (seven.JSONDecodeError, check.CheckError):
+                logging.warning("Could not parse event record id `%s`.", record_id)
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/storage/event_log/sql_event_log.py:979]: Potential subtle behavior change. Separate PR.

**Review by schrockn — CHANGES_REQUESTED** [review]
This PR is far too risky. It introduces code changes across the entire code base––sometimes it core infra. I've flagged a few of the file changes that IMO need be to addressed separately.

File: python_modules/dagster/dagster/_core/definitions/asset_reconciliation_sensor.py (hunk 5 above)
Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/asset_reconciliation_sensor.py:569]: Separate PR and have him review

**Comment by smackesey** [comment]
I believe I've separated out all the individual changes you flagged, so re-requesting review.

File: python_modules/dagster/dagster/_check/__init__.py [hunk 8]
```diff
@@ -167,13 +167,15 @@ def is_callable(obj: object, additional_message: Optional[str] = None) -> Callab
 # ##### CLASS
 # ########################
 
+T_Type = TypeVar("T_Type", bound=type)
+
 
 def class_param(
-    obj: object,
+    obj: T_Type,
```
Comment (schrockn) [inline python_modules/dagster/dagster/_check/__init__.py:174]: Not a huge fan of including this change until the discussion at https://github.com/dagster-io/internal/discussions/4618 has been resolved.

File: python_modules/dagster/dagster/_config/config_type.py [hunk 9]
```diff
@@ -404,7 +404,9 @@ def __init__(
     ):
         from .field import resolve_to_config_type
 
-        self.scalar_type = resolve_to_config_type(scalar_type)
+        self.scalar_type = check.inst(
+            cast(ConfigType, resolve_to_config_type(scalar_type)), ConfigType
```
Comment (schrockn) [inline python_modules/dagster/dagster/_config/config_type.py:408]: FWIW this cast here certainly convinced me that the "type verifying" approach is boilerplate-y and pedantic.

File: python_modules/dagster/dagster/_core/definitions/dependency.py [hunk 10]
```diff
@@ -505,14 +505,16 @@ def from_dict(cls, dict_repr: Dict[str, Any]) -> Optional["NodeHandle"]:
...
                 {
                     "name": dict_repr["parent"][0],
                     "parent": dict_repr["parent"][1],
                 }
             )
+        else:
+            parent = dict_repr["parent"]
 
-        return NodeHandle(**{k: dict_repr[k] for k in ["name", "parent"]})
+        return NodeHandle(name=dict_repr["name"], parent=parent)
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/dependency.py:517]: Not going to request changes based on it, but want to flag code changes that I wouldn't have done and the omitting would have made approve this PR a braindead rubberstamp. So this one is one.

File: python_modules/dagster/dagster/_core/definitions/freshness_policy.py [hunk 11]
```diff
@@ -173,8 +173,10 @@ def minutes_late(
                 self.cron_schedule, evaluation_time, ret_type=datetime.datetime, is_prev=True
             )
             evaluation_tick = next(schedule_ticks)
-        else:
+        elif evaluation_time is not None:
             evaluation_tick = evaluation_time
+        else:
+            check.failed("Must provide an evaluation time if not using a cron schedule")
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/freshness_policy.py:179]: Here is another one I would have deferred.

File: python_modules/dagster/dagster/_core/definitions/op_definition.py [hunk 12]
```diff
@@ -469,15 +467,15 @@ def _resolve_output_defs_from_outs(
         raise DagsterInvariantViolationError(
             "Expected Tuple annotation for multiple outputs, but received non-tuple annotation."
         )
-    if annotation != inspect.Parameter.empty and not len(annotation.__args__) == len(outs):
+    if annotation != inspect.Parameter.empty and not len(get_args(annotation)) == len(outs):
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/op_definition.py:470]: Here is another one I would have omitted.

File: python_modules/dagster/dagster/_core/definitions/partition.py [hunk 13]
```diff
@@ -196,10 +199,12 @@ def delta(self):
         else:
             check.failed(f"Unexpected ScheduleType {self}")
 
-    def __gt__(self, other):
+    def __gt__(self, other: "ScheduleType") -> bool:
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/partition.py:202]: seems slightly risk. same category as before.

File: python_modules/dagster/dagster/_core/definitions/resource_invocation.py [hunk 14]
```diff
@@ -84,14 +84,14 @@ def _check_invocation_requirements(
 
     # Construct a context if None was provided. This will initialize an ephemeral instance, and
     # console log manager.
-    init_context = init_context or build_init_resource_context()
+    _init_context = init_context or build_init_resource_context()
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/resource_invocation.py:87]: not your problem, but i wish we had a suitable naming convention for this type variable where you reassign it with a default valiue.

File: python_modules/dagster/dagster/_core/execution/plan/inputs.py [hunk 15]
```diff
@@ -219,14 +220,15 @@ def compute_version(self, step_versions, pipeline_def, resolved_run_config) -> O
...
-            io_manager_key
+            input_manager_key
         ]
 
-        op_config = resolved_run_config.solids.get(op.name)
+        op_config = check.not_none(resolved_run_config.solids.get(op.name))  # type: ignore  # fmt: skip
         input_config = op_config.inputs.get(self.input_name)
-        resource_config = resolved_run_config.resources.get(io_manager_key).config
+        resource_entry = check.not_none(resolved_run_config.resources.get(input_manager_key))  # type: ignore  # fmt: skip
+        resource_config = resource_entry.config
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/execution/plan/inputs.py:231]: another set of changes I would have omitted

File: python_modules/dagster/dagster/_core/selector/subset_selector.py [hunk 16]
```diff
@@ -35,8 +37,9 @@
 
 MAX_NUM = sys.maxsize
 
-T = TypeVar("T")
-DependencyGraph: TypeAlias = Mapping[str, Mapping[T, AbstractSet[T]]]
+T_Hashable = TypeVar("T_Hashable", bound=Hashable)
+Direction: TypeAlias = Literal["downstream", "upstream"]
+DependencyGraph: TypeAlias = Mapping[Direction, Mapping[T_Hashable, AbstractSet[T_Hashable]]]
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/selector/subset_selector.py:42]: even thought this is a type-only change, it is complex enough and affects mental model potentially and I would do it separately

File: python_modules/dagster/dagster/_daemon/run_coordinator/queued_run_coordinator_daemon.py [hunk 17]
```diff
@@ -194,14 +195,13 @@ def _get_runs_to_dequeue(
...
+        max_runs_to_launch = max_concurrent_runs - len_iter(in_progress_runs)
         if max_concurrent_runs_enabled:
-            max_runs_to_launch = max_concurrent_runs - len(in_progress_runs)
-
             # Possibly under 0 if runs were launched without queuing
             if max_runs_to_launch <= 0:
                 self._logger.info(
                     "{} runs are currently in progress. Maximum is {}, won't launch more.".format(
-                        len(in_progress_runs), max_concurrent_runs
+                        len_iter(in_progress_runs), max_concurrent_runs
```
Comment (schrockn) [inline python_modules/dagster/dagster/_daemon/run_coordinator/queued_run_coordinator_daemon.py:204]: another thing I would address separately (maybe in a follow up). I'd love to have some agreed upon norms to handle `Mapping`, `Sequence`, and `Iterable` classes, because a lot of the idioms right now are annoying. For example this just came up yesterday https://github.com/dagster-io/dagster/pull/11713#discussion_r1070198129

**Review by schrockn — APPROVED** [review]
Ok this is good to go. I made comments about changes I would have but into a sep PR. In particular I think you should made one change that was typing only the packaged up all changes that have _zero_ runtime risk, and then one change that dealt with all the changes that might even remotely have runtime behavior changes, which would have made this much easier to review and therefore less risky.

File: python_modules/dagster/dagster/_daemon/run_coordinator/queued_run_coordinator_daemon.py (hunk 17 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_daemon/run_coordinator/queued_run_coordinator_daemon.py:204]: My approach is:

- Use `Mapping` everywhere for dict unless we intend to mutate the dict.
- Use `Sequence` everywhere for list unless we intend to mutate the list.
- `Iterable` doesn't come up frequently, I've assumed when it does appear it's because we have an abstract method that might return either a list or an iterator. It's possible there are some places in the code where `Iterable` should instead be sequence (like in this case, where we are checking length).
